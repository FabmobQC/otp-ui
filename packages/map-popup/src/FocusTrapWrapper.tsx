import React, { useCallback } from "react";

/**
 * Helper method to find the element within the app menu at the given offset
 * (e.g. previous or next) relative to the specified element.
 *
 * @param {string} query  - Argument that gets passed to document.querySelectorAll
 * @param {HTMLElement} element - Specified element (e.target)
 * @param {1 | -1} offset - Determines direction to move within array of focusable elements (previous or next)
 * @returns {HTMLElement} - element to be focused
 */

export function getEntryRelativeTo(
  query: string,
  element: EventTarget,
  offset: 1 | -1
): HTMLElement {
  const entries = Array.from(document.querySelectorAll(query));
  const firstElement = entries[0];
  const lastElement = entries[entries.length - 1];
  const elementIndex = entries.indexOf(element as HTMLButtonElement);

  if (element === firstElement && offset === -1) {
    return lastElement as HTMLElement;
  }
  if (element === lastElement && offset === 1) {
    return firstElement as HTMLElement;
  }
  return entries[elementIndex + offset] as HTMLElement;
}

const FocusTrapWrapper = ({
  children,
  id,
  setPopup
}: {
  children: JSX.Element | JSX.Element[];
  id: string;
  setPopup: any;
}): JSX.Element => {
  const queryId = `${id}-popup-focus-trap button`;
  const handleKeyDown = useCallback(
    e => {
      const element = e.target as HTMLElement;
      switch (e.key) {
        case "Escape":
          setPopup(false);
          break;
        case "Tab":
          if (e.shiftKey) {
            e.preventDefault();
            getEntryRelativeTo(queryId, element, -1)?.focus();
          } else {
            e.preventDefault();
            getEntryRelativeTo(queryId, element, 1)?.focus();
          }
          break;
        default:
      }
    },
    [setPopup]
  );

  return (
    <div
      id={`${id}-popup-focus-trap`}
      onKeyDown={handleKeyDown}
      role="presentation"
    >
      {children}
    </div>
  );
};

export default FocusTrapWrapper;
