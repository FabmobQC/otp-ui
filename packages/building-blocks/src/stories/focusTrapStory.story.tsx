import React from "react";
import FocusTrapWrapper from "../focus-trap-wrapper";

export default {
  title: "Building-Blocks/FocusTrapWrapper",
  component: FocusTrapWrapper
};

export const FocusTrapAroundButtonSet = (): React.ReactElement => (
  <FocusTrapWrapper id="button-set-story">
    <button type="button">Button 1</button>
    <button type="button">Button 2</button>
    <button type="button">Button 3</button>
    <button type="button">Button 4</button>
  </FocusTrapWrapper>
);

export const FocusTrapWithVariousEls = (): React.ReactElement => (
  <FocusTrapWrapper
    id="various-els-story"
    focusElements={["button", "a", "div", "input", "select"]}
  >
    <button type="button">Button 1</button>
    <a href="/">link</a>
    <div tabIndex={-1}>focusable div</div>
    <input disabled type="text"></input>
    <input type="text"></input>
    <select>
      <option>Option 1</option>
      <option>Option 2</option>
    </select>
  </FocusTrapWrapper>
);
