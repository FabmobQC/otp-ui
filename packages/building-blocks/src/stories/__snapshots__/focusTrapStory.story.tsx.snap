// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Building-Blocks/FocusTrapWrapper FocusTrapAroundButtonSet smoke-test 1`] = `
<div id="button-set-story-focus-trap"
     role="presentation"
>
  <button type="button">
    Button 1
  </button>
  <button type="button">
    Button 2
  </button>
  <button type="button">
    Button 3
  </button>
  <button type="button">
    Button 4
  </button>
</div>
`;

exports[`Building-Blocks/FocusTrapWrapper FocusTrapWithVariousEls smoke-test 1`] = `
<div id="various-els-story-focus-trap"
     role="presentation"
>
  <button type="button">
    Button 1
  </button>
  <a href="/">
    link
  </a>
  <div tabindex="-1">
    focusable div
  </div>
  <input disabled
         type="text"
  >
  <input type="text">
  <select>
    <option>
      Option 1
    </option>
    <option>
      Option 2
    </option>
  </select>
</div>
`;
